.gitlab-ci.yml

```diff
@@ -39,18 +206,120 @@ nix-dry:
     - nix-build -v -v --dry-run ./release.nix --attr package.macos.x64.macho
 
 nix:
+  image: registry.gitlab.com/matrixai/engineering/maintenance/gitlab-runner
   stage: build
   interruptible: true
   script:
     - >
-        nix-build ./release.nix
-        --max-jobs $(nproc)
-        --attr application
-        --attr docker
+      BUILDOUT=`nix-build ./release.nix\
+        --attr application\
+        --attr docker\
+        --attr package.linux.x64.elf\
+        --attr package.windows.x64.exe\
+        --attr package.macos.x64.macho`;
+        mkdir builds;
+        cp -r $BUILDOUT builds/;
+        chmod -R a+rw builds;
   only:
     - master
+  artifacts:
+    paths:
+      - builds
+
+runs in node:
+  image: registry.gitlab.com/matrixai/engineering/maintenance/gitlab-runner
+  stage: quality
+  only:
+    - master
+  interruptible: true
+  script:
+    - >
+      nix-shell -I nixpkgs=./pkgs.nix --packages nodejs --run '
+      npm install;
+      echo password > testPassword;
+      npm run polykey -- agent start --test -np ./nodePath --password-file ./testPassword;
+      '
+
+running application build:
+  image: registry.gitlab.com/matrixai/engineering/maintenance/gitlab-runner
+  stage: quality
+  only:
+    - master
+  needs:
+    - job: nix
+      artifacts: true
+  interruptible: true
+  script:
+    - >
+      nix-shell -I nixpkgs=./pkgs.nix --packages nodejs --run '
+      echo password > testPassword;
+      node builds/*/lib/node_modules/@matrixai/polykey/dist/bin/polykey.js agent start -np ./nodePath --password-file ./testPassword --test;
+      '
+
+
+running docker build:
+  image: docker:19.03.12
+  variables:
+    DOCKER_TLS_CERTDIR: "/certs"
+  services:
+    - docker:19.03.12-dind
+  before_script:
+    - docker info
+  stage: quality
+  only:
+    - master
+  needs:
+    - job: nix
+      artifacts: true
+  interruptible: true
+  script:
+    - ./scripts/runDocker.sh
+
+running linux build:
+  stage: quality
+  only:
+    - master
+  tags:
+    - linux
```

Comment:
Should investigate whether specifying a single container image like Alpine linux is better than just saying any linux image.

Reply:
I'm looking into this but it's hard to find useful information relating to this.
We can specify the image so we have a linux runner but would this be a VM or a docker image? how do we specify a VM image? how do we find and specify a windows image?

so far as I can find the tags are the recommended method but this information for this use case is light. so far the tags have worked fine and I think should continue to work for or purposes.

Reply:
Linux would always use a container image. In fact the usage of `image` option at the root of the `.gitlab-ci.yml` file means that the container image is used by default across all jobs.

Reply:
See: https://docs.gitlab.com/ee/ci/quick_start/#gitlab-ciyml-tips and https://docs.gitlab.com/ee/ci/yaml/index.html

Reply:
I believe you shouldn't even need to specify `tags: linux` since it should be the default runner being used. Plus https://docs.gitlab.com/ee/ci/runners/runner_cloud/linux_runner_cloud.html this indicates that `image` is always used by the linux runners, which means it always means a container image.

This basically means that our `.gitlab-ci.yml` file has to change the `image` setting. For some of the jobs, we can put `image: registry.gitlab.com/matrixai/engineering/maintenance/gitlab-runner` which will use our runner. This should be used for `lint`, `test` and `nix-dry`, `nix`, `packages` as usual. But your new jobs will need to use `image: alpine:3.15` or `alpine:latest`.

Then for your windows and mac runners, the `image` probably means VM and not container.

Reply:
For windows: https://docs.gitlab.com/ee/ci/runners/runner_cloud/windows_runner_cloud.html

> The Windows runner virtual machine instances do not use the GitLab Docker executor. This means that you can’t specify image or services in your pipeline configuration.

Available software already installed on windows are documented here: https://gitlab.com/gitlab-org/ci-cd/shared-runners/images/gcp/windows-containers/blob/main/cookbooks/preinstalled-software/README.md

This means Windows runners are natively executing the code, no VM and no containers are used.

In that list it shows that chocolatey exists, that would allow us to install other windows software. But think of it as a base windows install, and therefore our single executable should just work there.

Reply:
And for windows, the `image` is specified. But it means the VM instead of container: https://docs.gitlab.com/ee/ci/runners/runner_cloud/macos/environment.html#vm-images

Reply:
This has been resolved.